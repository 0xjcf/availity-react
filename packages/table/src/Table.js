import React from 'react';
import 'core-js/stable';
import PropTypes from 'prop-types';
import { useTable, useRowSelect, useSortBy } from 'react-table';
import { Util, Table as RsTable } from 'reactstrap';
import Icon from '@availity/icon';
import TableHeader from './TableHeader';
import TableHeaderRow from './TableHeaderRow';
import TableHeaderCell from './TableHeaderCell';
import IndeterminateCheckbox from './IndeterminateCheckbox';
import TableRow from './TableRow';
import TableCell from './TableCell';
import { TableContext } from './TableContext';

const Table = ({
  id,
  additionalContent: AdditionalContent,
  bodyProps,
  columns,
  cellProps,
  headerProps,
  onRowClick,
  onRowSelected,
  records,
  rowProps,
  scrollable,
  selectable,
  sortable,
  initialState,
  ...rest
}) => {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    selectedFlatRows: selectedRows,
    allColumns,
    toggleAllRowsSelected,
  } = useTable(
    {
      columns,
      data: records,
      initialState: initialState || {},
    },
    useSortBy,
    useRowSelect,
    (hooks) => {
      if (selectable) {
        hooks.visibleColumns.push((columns) => [
          {
            id: 'selection',
            title: 'Select record(s)',
            className: 'fixed-width-selection',
            defaultCanSort: false,
            disableClick: true,
            Header: ({ getToggleAllRowsSelectedProps }) => {
              const selectedProps = {
                onClick: async () => {
                  await toggleAllRowsSelected();
                  if (onRowSelected) {
                    onRowSelected({ selectedRows });
                  }
                },
              };

              return (
                <div className="text-center">
                  <IndeterminateCheckbox
                    data-testid="table_header_select_all"
                    aria-label="Select all records"
                    {...selectedProps}
                    {...getToggleAllRowsSelectedProps()}
                  />
                </div>
              );
            },
            Cell: ({ row: { original, toggleRowSelected, getToggleRowSelectedProps, index } }) => {
              const selectedProps = {
                onClick: () => {
                  if (onRowSelected) {
                    toggleRowSelected();
                    onRowSelected({ selectedId: original.id, data: original });
                  }
                },
              };

              return (
                <div className="text-center">
                  <IndeterminateCheckbox
                    data-testid={`table_header_select_row_${index}`}
                    aria-label="Select record"
                    {...selectedProps}
                    {...getToggleRowSelectedProps()}
                  />
                </div>
              );
            },
          },
          ...columns,
        ]);
      }

      if (!sortable) {
        hooks.visibleColumns.forEach((col) => {
          col.disableSortBy = true;
        });
      }
    }
  );

  let handleRowClick;
  let handleCellClick;

  if (selectable && onRowClick) {
    handleCellClick = onRowClick;
  } else {
    handleRowClick = onRowClick;
  }

  const populateId = () => (id ? `${id}_` : '');

  return (
    <TableContext.Provider
      value={{
        allColumns,
        selectedRows,
        scrollable,
        AdditionalContent,
      }}
    >
      <RsTable id={id} {...getTableProps({ className: 'av-grid' })} {...rest}>
        <TableHeader id={`${populateId()}table_header`} {...headerProps}>
          {headerGroups.map((headerGroup, rowIndex) => (
            <TableHeaderRow
              id={`${populateId()}table_header_row_${rowIndex}`}
              data-testid={`${populateId()}table_header_row_${rowIndex}`}
              key={rowIndex.toString()}
              headerGroup={headerGroup}
            >
              {headerGroup.headers.map((column, cellIndex) => (
                <TableHeaderCell
                  id={`${populateId()}table_header_row_${rowIndex}_cell_${cellIndex}_${column.id}`}
                  data-testid={`${populateId()}table_header_row_${rowIndex}_cell_${cellIndex}_${column.id}`}
                  key={column.id}
                  column={column}
                >
                  {column.render('Header')}
                  {sortable && column.defaultCanSort ? (
                    <Icon
                      aria-hidden="true"
                      name={column.isSorted ? (column.isSortedDesc ? 'sort-down' : 'sort-up') : 'sort'}
                    />
                  ) : null}
                </TableHeaderCell>
              ))}
            </TableHeaderRow>
          ))}
        </TableHeader>
        <tbody {...getTableBodyProps()} {...bodyProps}>
          {rows.map((row, rowIndex) => {
            prepareRow(row);
            return (
              <TableRow
                id={`${populateId()}table_row_${rowIndex}`}
                data-testid={`${populateId()}table_row_${rowIndex}`}
                key={row.original.id}
                index={rowIndex}
                row={row}
                onRowClick={handleRowClick}
                onCellClick={handleCellClick}
                {...rowProps}
              >
                {row.cells.map((cell, cellIndex) => (
                  <TableCell
                    id={`${populateId()}table_row_${rowIndex}_cell_${cellIndex}`}
                    data-testid={`${populateId()}table_row_${rowIndex}_cell_${cellIndex}`}
                    key={cellIndex.toString()}
                    cell={cell}
                    onCellClick={handleCellClick}
                    {...cellProps}
                  >
                    {cell.render('Cell')}
                  </TableCell>
                ))}
              </TableRow>
            );
          })}
        </tbody>
      </RsTable>
    </TableContext.Provider>
  );
};

Table.propTypes = {
  id: PropTypes.string,
  additionalContent: Util.tagPropType,
  bodyProps: PropTypes.object,
  cellProps: PropTypes.object,
  columns: PropTypes.arrayOf(PropTypes.object),
  headerProps: PropTypes.object,
  initialState: PropTypes.object,
  onRowClick: PropTypes.func,
  onRowSelected: PropTypes.func,
  selectable: PropTypes.bool,
  sortable: PropTypes.bool,
  scrollable: PropTypes.bool,
  records: PropTypes.arrayOf(PropTypes.object),
  rowProps: PropTypes.object,
  sortBy: PropTypes.arrayOf(PropTypes.object),
  getToggleAllRowsSelectedProps: PropTypes.func,
  row: PropTypes.object,
};

Table.defaultProps = {
  headerProps: {},
  rowProps: {},
  bodyProps: {},
  cellProps: {},
};

export default Table;
