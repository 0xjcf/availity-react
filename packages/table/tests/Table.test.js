import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import Table, { ActionCell, BadgeCell, CurrencyCell, IconCell, DateCell } from '..';
import basicData from './data/basicData.json';
import formattedData from './data/needsFormattedData.json';

const basicColumns = [
  {
    Header: 'First Name',
    accessor: 'first_name',
  },
  {
    Header: 'Last Name',
    accessor: 'last_name',
  },
  {
    Header: 'Email',
    accessor: 'email',
  },
];

const formattedColumns = [
  {
    Header: 'Basic Text',
    accessor: 'text',
  },
  {
    Header: 'Currency',
    accessor: 'currency',
    Cell: CurrencyCell({}),
  },
  {
    Header: 'Badge',
    accessor: 'badge',
    Cell: ({ row: { original } }) => (original ? BadgeCell('badge-success', original.badge) : null),
  },
  {
    Header: 'Icon',
    accessor: 'icon',
    Cell: IconCell({ name: 'doc-alt', title: 'View Notes' }),
  },
  {
    Header: 'Date',
    accessor: 'date',
    Cell: DateCell({ dateFormat: 'MM/DD/yyyy' }),
  },
  {
    id: 'actions',
    Header: 'Actions',
    canSort: false,
    Cell: ActionCell({
      actions: [
        {
          id: 'action1',
          displayText: 'Action 1',
          isVisible: (record) => record.icon,
          onClick: (record) => {
            // eslint-disable-next-line no-console
            console.log(`action on record ${record.id}`);
          },
        },
        {
          id: 'actionDivider',
          divider: true,
        },
        {
          id: 'action2',
          displayText: 'Action 2',
          onClick: (record) => {
            // eslint-disable-next-line no-console
            console.log(`action on record ${record.id}`);
          },
        },
      ],
    }),
  },
];

describe('Table', () => {
  test('should render basic table', () => {
    const { container } = render(<Table records={basicData} columns={basicColumns} />);

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render selectable table', async () => {
    const onSelect = jest.fn();
    const { container, getByTestId } = render(
      <Table selectable records={basicData} columns={basicColumns} onRowSelected={onSelect} />
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const selectAllRecordsInput = getByTestId('table_header_select_all');
    fireEvent.click(selectAllRecordsInput);

    await waitFor(() => {
      expect(onSelect).toHaveBeenCalledTimes(2);
    });

    const selectFirstRecordInput = getByTestId('table_header_select_row_0');
    fireEvent.click(selectFirstRecordInput);

    await waitFor(() => {
      expect(onSelect).toHaveBeenCalledTimes(3);
    });
  });

  test('should render sortable table', () => {
    const { container } = render(<Table sortable records={basicData} columns={basicColumns} />);

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render with formatted cells', () => {
    const { container } = render(<Table records={formattedData} columns={formattedColumns} />);

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render with expected ids', () => {
    const { container } = render(<Table id="my_availity_table" records={basicData} columns={basicColumns} />);

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should call onRowClick when event is provided', async () => {
    const onRowClick = jest.fn();
    const { container, getByTestId } = render(
      <Table records={basicData} columns={basicColumns} onRowClick={onRowClick} />
    );

    expect(container).toBeDefined();

    const tableRow = getByTestId('table_row_0');

    fireEvent.click(tableRow);

    await waitFor(() => {
      expect(onRowClick).toHaveBeenCalledTimes(1);
    });
  });

  test('should call onRowClick when event is provided for selectable table', async () => {
    const onRowClick = jest.fn();
    const { container, getByTestId } = render(
      <Table selectable records={basicData} columns={basicColumns} onRowClick={onRowClick} />
    );

    expect(container).toBeDefined();

    const tableRow = getByTestId('table_row_0_cell_1');

    fireEvent.click(tableRow);

    await waitFor(() => {
      expect(onRowClick).toHaveBeenCalledTimes(1);
    });
  });
});
