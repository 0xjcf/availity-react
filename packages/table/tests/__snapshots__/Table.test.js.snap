// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render basic table 1`] = `
<div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          style="cursor: pointer;"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          style="cursor: pointer;"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          style="cursor: pointer;"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
          role="cell"
          title="Barbra"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
          role="cell"
          title="Tull"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
          role="cell"
          title="btull0@usatoday.com"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
          role="cell"
          title="Karlis"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
          role="cell"
          title="Yurshev"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
          role="cell"
          title="kyurshev1@spotify.com"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
          role="cell"
          title="Bordie"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
          role="cell"
          title="Marsy"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
          role="cell"
          title="bmarsy2@elegantthemes.com"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render selectable table 1`] = `
<div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          class="fixed-width-selection"
          colspan="1"
          data-testid="table_header_row_0_cell_0_selection"
          id="table_header_row_0_cell_0_selection"
          role="columnheader"
        >
          <div
            class="text-center"
          >
            <label
              class="custom-control custom-checkbox"
            >
              <input
                aria-label="Select all records"
                class="custom-control-input"
                data-testid="table_header_select_all"
                style="cursor: pointer;"
                title="Toggle All Rows Selected"
                type="checkbox"
              />
              <span
                class="custom-control-label"
              />
            </label>
          </div>
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_first_name"
          id="table_header_row_0_cell_1_first_name"
          role="columnheader"
          style="cursor: pointer;"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_last_name"
          id="table_header_row_0_cell_2_last_name"
          role="columnheader"
          style="cursor: pointer;"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_3_email"
          id="table_header_row_0_cell_3_email"
          role="columnheader"
          style="cursor: pointer;"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          class="fixed-width-selection"
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
          role="cell"
        >
          <div
            class="text-center"
          >
            <label
              class="custom-control custom-checkbox"
            >
              <input
                aria-label="Select record"
                class="custom-control-input"
                data-testid="table_header_select_row_0"
                style="cursor: pointer;"
                title="Toggle Row Selected"
                type="checkbox"
              />
              <span
                class="custom-control-label"
              />
            </label>
          </div>
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
          role="cell"
          title="Barbra"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
          role="cell"
          title="Tull"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_3"
          id="table_row_0_cell_3"
          role="cell"
          title="btull0@usatoday.com"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          class="fixed-width-selection"
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
          role="cell"
        >
          <div
            class="text-center"
          >
            <label
              class="custom-control custom-checkbox"
            >
              <input
                aria-label="Select record"
                class="custom-control-input"
                data-testid="table_header_select_row_1"
                style="cursor: pointer;"
                title="Toggle Row Selected"
                type="checkbox"
              />
              <span
                class="custom-control-label"
              />
            </label>
          </div>
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
          role="cell"
          title="Karlis"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
          role="cell"
          title="Yurshev"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_3"
          id="table_row_1_cell_3"
          role="cell"
          title="kyurshev1@spotify.com"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          class="fixed-width-selection"
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
          role="cell"
        >
          <div
            class="text-center"
          >
            <label
              class="custom-control custom-checkbox"
            >
              <input
                aria-label="Select record"
                class="custom-control-input"
                data-testid="table_header_select_row_2"
                style="cursor: pointer;"
                title="Toggle Row Selected"
                type="checkbox"
              />
              <span
                class="custom-control-label"
              />
            </label>
          </div>
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
          role="cell"
          title="Bordie"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
          role="cell"
          title="Marsy"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_3"
          id="table_row_2_cell_3"
          role="cell"
          title="bmarsy2@elegantthemes.com"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render sortable table 1`] = `
<div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          style="cursor: pointer;"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          style="cursor: pointer;"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          style="cursor: pointer;"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
          role="cell"
          title="Barbra"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
          role="cell"
          title="Tull"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
          role="cell"
          title="btull0@usatoday.com"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
          role="cell"
          title="Karlis"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
          role="cell"
          title="Yurshev"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
          role="cell"
          title="kyurshev1@spotify.com"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
          role="cell"
          title="Bordie"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
          role="cell"
          title="Marsy"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
          role="cell"
          title="bmarsy2@elegantthemes.com"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render with expected ids 1`] = `
<div>
  <table
    class="av-grid table"
    id="my_availity_table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="my_availity_table_table_header"
    >
      <tr
        data-testid="my_availity_table_table_header_row_0"
        id="my_availity_table_table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="my_availity_table_table_header_row_0_cell_0_first_name"
          id="my_availity_table_table_header_row_0_cell_0_first_name"
          role="columnheader"
          style="cursor: pointer;"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="my_availity_table_table_header_row_0_cell_1_last_name"
          id="my_availity_table_table_header_row_0_cell_1_last_name"
          role="columnheader"
          style="cursor: pointer;"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="my_availity_table_table_header_row_0_cell_2_email"
          id="my_availity_table_table_header_row_0_cell_2_email"
          role="columnheader"
          style="cursor: pointer;"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="my_availity_table_table_row_0"
        id="my_availity_table_table_row_0"
        role="row"
      >
        <td
          data-testid="my_availity_table_table_row_0_cell_0"
          id="my_availity_table_table_row_0_cell_0"
          role="cell"
          title="Barbra"
        >
          Barbra
        </td>
        <td
          data-testid="my_availity_table_table_row_0_cell_1"
          id="my_availity_table_table_row_0_cell_1"
          role="cell"
          title="Tull"
        >
          Tull
        </td>
        <td
          data-testid="my_availity_table_table_row_0_cell_2"
          id="my_availity_table_table_row_0_cell_2"
          role="cell"
          title="btull0@usatoday.com"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="my_availity_table_table_row_1"
        id="my_availity_table_table_row_1"
        role="row"
      >
        <td
          data-testid="my_availity_table_table_row_1_cell_0"
          id="my_availity_table_table_row_1_cell_0"
          role="cell"
          title="Karlis"
        >
          Karlis
        </td>
        <td
          data-testid="my_availity_table_table_row_1_cell_1"
          id="my_availity_table_table_row_1_cell_1"
          role="cell"
          title="Yurshev"
        >
          Yurshev
        </td>
        <td
          data-testid="my_availity_table_table_row_1_cell_2"
          id="my_availity_table_table_row_1_cell_2"
          role="cell"
          title="kyurshev1@spotify.com"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="my_availity_table_table_row_2"
        id="my_availity_table_table_row_2"
        role="row"
      >
        <td
          data-testid="my_availity_table_table_row_2_cell_0"
          id="my_availity_table_table_row_2_cell_0"
          role="cell"
          title="Bordie"
        >
          Bordie
        </td>
        <td
          data-testid="my_availity_table_table_row_2_cell_1"
          id="my_availity_table_table_row_2_cell_1"
          role="cell"
          title="Marsy"
        >
          Marsy
        </td>
        <td
          data-testid="my_availity_table_table_row_2_cell_2"
          id="my_availity_table_table_row_2_cell_2"
          role="cell"
          title="bmarsy2@elegantthemes.com"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render with formatted cells 1`] = `
<div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_text"
          id="table_header_row_0_cell_0_text"
          role="columnheader"
          style="cursor: pointer;"
          title="Basic Text"
        >
          Basic Text
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_currency"
          id="table_header_row_0_cell_1_currency"
          role="columnheader"
          style="cursor: pointer;"
          title="Currency"
        >
          Currency
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_badge"
          id="table_header_row_0_cell_2_badge"
          role="columnheader"
          style="cursor: pointer;"
          title="Badge"
        >
          Badge
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_3_icon"
          id="table_header_row_0_cell_3_icon"
          role="columnheader"
          style="cursor: pointer;"
          title="Icon"
        >
          Icon
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_4_date"
          id="table_header_row_0_cell_4_date"
          role="columnheader"
          style="cursor: pointer;"
          title="Date"
        >
          Date
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_5_actions"
          id="table_header_row_0_cell_5_actions"
          role="columnheader"
          title="Actions"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
          role="cell"
          title="My Text Field"
        >
          My Text Field
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
          role="cell"
          title="43.08"
        >
          <span
            title="$43.08"
          >
            $43.08
          </span>
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
          role="cell"
          title="badge field"
        >
          <p
            class="badge badge-success"
            title="badge field"
          >
            badge field
          </p>
        </td>
        <td
          data-testid="table_row_0_cell_3"
          id="table_row_0_cell_3"
          role="cell"
          title="true"
        >
          <i
            aria-hidden="true"
            class="icon icon-doc-alt"
            title="View Notes"
          />
        </td>
        <td
          data-testid="table_row_0_cell_4"
          id="table_row_0_cell_4"
          role="cell"
          title="1956-02-16T05:00:00.000+0000"
        >
          <span
            title="02/16/1956"
          >
            02/16/1956
          </span>
        </td>
        <td
          data-testid="table_row_0_cell_5"
          id="table_row_0_cell_5"
          role="cell"
        >
          <div
            class="dropdown-action-menu dropleft dropdown"
            id="table_row_action_menu_1"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="btn btn-ghost btn btn-secondary"
              data-boundary="viewport"
              id="table_row_action_menu_1_dropdown_toggle"
              type="button"
            >
              <i
                class="icon icon-menu"
                id="table_row_action_menu_1_dropdown_toggle_icon"
              />
            </button>
            <div
              aria-hidden="true"
              class="dropdown-action-menu dropdown-menu"
              id="table_row_action_menu_1_dropdown_toggle_icon"
              role="menu"
              tabindex="-1"
            >
              <button
                class="dropdown-item"
                id="table_row_action_menu_item_0_action_action1"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Action 1
              </button>
              <button
                class="dropdown-item"
                id="table_row_action_menu_item_1_action_action2"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Action 2
              </button>
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
          role="cell"
          title="My Text Field"
        >
          My Text Field
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
          role="cell"
          title="43.08"
        >
          <span
            title="$43.08"
          >
            $43.08
          </span>
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
          role="cell"
          title="badge field"
        >
          <p
            class="badge badge-success"
            title="badge field"
          >
            badge field
          </p>
        </td>
        <td
          data-testid="table_row_1_cell_3"
          id="table_row_1_cell_3"
          role="cell"
        />
        <td
          data-testid="table_row_1_cell_4"
          id="table_row_1_cell_4"
          role="cell"
          title="1990-02-16T05:00:00.000+0000"
        >
          <span
            title="02/16/1990"
          >
            02/16/1990
          </span>
        </td>
        <td
          data-testid="table_row_1_cell_5"
          id="table_row_1_cell_5"
          role="cell"
        >
          <div
            class="dropdown-action-menu dropleft dropdown"
            id="table_row_action_menu_2"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="btn btn-ghost btn btn-secondary"
              data-boundary="viewport"
              id="table_row_action_menu_2_dropdown_toggle"
              type="button"
            >
              <i
                class="icon icon-menu"
                id="table_row_action_menu_2_dropdown_toggle_icon"
              />
            </button>
            <div
              aria-hidden="true"
              class="dropdown-action-menu dropdown-menu"
              id="table_row_action_menu_2_dropdown_toggle_icon"
              role="menu"
              tabindex="-1"
            >
              <button
                class="dropdown-item"
                id="table_row_action_menu_item_0_action_action1"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Action 1
              </button>
              <button
                class="dropdown-item"
                id="table_row_action_menu_item_1_action_action2"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Action 2
              </button>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
